src/Execution/MutationExecutor.php

```diff
@@ -270,6 +307,101 @@ public static function executeUpdate(Model $model, Collection $args, ?Relation $
 
         $model = $model->newQuery()->findOrFail($id);
 
+        return self::executeUpdateWithLoadedModel($model, $args, $parentRelation);
+    }
+
+    /**
+     * Extract all the arguments that correspond to a relation of a certain type on the model.
+     *
+     * For example, if the args input looks like this:
+     *
+     * [
+     *  'comments' =>
+     *    ['foo' => 'Bar'],
+     *  'name' => 'Ralf',
+     * ]
+     *
+     * and the model has a method "comments" that returns a HasMany relationship,
+     * the result will be:
+     * [
+     *   [
+     *    'comments' =>
+     *      ['foo' => 'Bar'],
+     *   ],
+     *   [
+     *    'name' => 'Ralf',
+     *   ]
+     * ]
+     *
+     * @param  \ReflectionClass  $modelReflection
+     * @param  \Illuminate\Support\Collection  $args
+     * @param  string  $relationClass
+     * @return \Illuminate\Support\Collection  [relationshipArgs, remainingArgs]
+     */
+    protected static function partitionArgsByRelationType(ReflectionClass $modelReflection, Collection $args, string $relationClass): Collection
+    {
+        return $args->partition(
+            function ($value, string $key) use ($modelReflection, $relationClass): bool {
+                if (! $modelReflection->hasMethod($key)) {
+                    return false;
+                }
+
+                $relationMethodCandidate = $modelReflection->getMethod($key);
+                if (! $returnType = $relationMethodCandidate->getReturnType()) {
+                    return false;
+                }
+
+                if (! $returnType instanceof ReflectionNamedType) {
+                    return false;
+                }
+
+                return is_a($returnType->getName(), $relationClass, true);
+            }
+        );
+    }
+
+    /**
+     * Execute an upsert mutation.
+     *
+     * @param \Illuminate\Database\Eloquent\Model $model
+     *         An empty instance of the model that should be created or updated
+     * @param \Illuminate\Support\Collection $args
+     *         The corresponding slice of the input arguments for creating or updating this model
+     * @param \Illuminate\Database\Eloquent\Relations\Relation|null $parentRelation
+     *         If we are in a nested upsert, we can use this to associate the new model to its parent
+     *
+     * @return \Illuminate\Database\Eloquent\Model
+     */
+    public static function executeUpsert(Model $model, Collection $args, ?Relation $parentRelation = null): Model
+    {
+        $id = $args->pull('id')
+            ?? $args->pull(
+                $model->getKeyName()
+            );
+
+        try {
+            $model = $model->newQuery()->findOrFail($id);
```

Comment:
Why not just just `find()` and check if a `$model` is returned?

Reply:
It is the same result. I used findOrFail because it is more redeable that a find with an `$model !== null` . It is much more clear to read `findOrFail`with a `ModelNotFound`. I really didn't do a performance test to see the difference in performance.

Reply:
Do you prefer find with `if $model !== null`? I can change it :-)

Reply:
`if($model){}` should suffice

Reply:
Just reread that part, i think we should rename to `$modelInstance` or something else, to differentiate from the parameter.